File: leetcode/207-course-schedule/main.py
"""
    1st approach: topo ordering BFS with array

    Time    O(V+E)
    Space   O(V)
    80 ms, faster than 60.39%
    7may2019
"""


class Solution(object):
    def canFinish(self, numCourses, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: bool
        """
        connections = []
        indegrees = []
        for _ in range(numCourses):
            indegrees.append(0)
            connections.append([])

        for cur, prev in prerequisites:
            indegrees[cur] += 1
            connections[prev].append(cur)

        q = []
        for i in range(len(indegrees)):
            if indegrees[i] == 0:
                q.append(i)

        res = []
        while len(q) > 0:
            pop0 = q.pop(0)
            res.append(pop0)
            children = connections[pop0]
            for child in children:
                indegrees[child] -= 1
                if indegrees[child] == 0:
                    q.append(child)
        return len(res) == numCourses


print(Solution().canFinish(2, [[1, 0]]))
print(Solution().canFinish(2, [[0, 1], [1, 0]]))
print(Solution().canFinish(
    6, [[4, 3], [1, 0], [5, 2], [5, 4], [5, 1], [2, 3]]))
print(Solution().canFinish(
    6, [[4, 3], [1, 0], [5, 2], [5, 4], [5, 1], [2, 3], [3, 5]]))
print("-----")

"""
    2nd approach: topo ordering BFS but with hashtable

    Time    O(V+E)
    Space   O(V)
    216 ms, faster than 20.38%
    28mar2019
"""


class Solution(object):
    def canFinish(self, numCourses, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: List[int]
        """
        n = numCourses
        graph = {} # key: []
        indegrees = {} # key: count
        for i in range(n):
            graph[i] = []
            indegrees[i] = 0
        for a, b in prerequisites:
            graph[b].append(a)
            indegrees[a] += 1
        q = []
        for cid in indegrees:
            if indegrees[cid] == 0:
                q.append(cid)
        visited = set()
        while len(q) > 0:
            cid = q.pop(0)
            visited.add(cid)
            for child in graph[cid]:
                indegrees[child] -= 1
                if indegrees[child] == 0:
                    q.append(child)
        return len(visited) == n


print(Solution().canFinish(2, [[1, 0]]))
print(Solution().canFinish(2, [[0, 1], [1, 0]]))
print(Solution().canFinish(
    6, [[4, 3], [1, 0], [5, 2], [5, 4], [5, 1], [2, 3]]))
print(Solution().canFinish(
    6, [[4, 3], [1, 0], [5, 2], [5, 4], [5, 1], [2, 3], [3, 5]]))

File: leetcode/207-course-schedule/main.go
package main

import "fmt"

/*
	1st approach got LTE in python
*/

/*
	2nd approach:

	!!! actually i was almost there in 1st approach!!!
	!!! this method is known as topological sorting/ordering !!!

	1. create a list to save to children for each node
	2. for each node, put the children in
			e.g. [4, 3], [1, 0], [5, 2], [5, 4], [5, 1], [2, 3]
			children list = [[], [0], [3], [], [3], [2,4,1]]
	3. dfs each node to see if there is a cycle
	4. return true if no cycle

	Time	O(V+E) V: number of vertexes, E: number og edges
	Space	O(V)

	16ms beats 94.81%
*/
func canFinish(numCourses int, prerequisites [][]int) bool {
	// prepare a list to save to children for each node
	graph := [][]int{}
	for i := 0; i < numCourses; i++ {
		graph = append(graph, []int{})
	}
	// iterate though the edges and put them into the corresponding nodes in the graph
	for i := 0; i < len(prerequisites); i++ {
		prereq := prerequisites[i]
		prev := prereq[0]
		cur := prereq[1]
		graph[prev] = append(graph[prev], cur)
	}
	// iterate though the nodes and see if there is a cycle
	seen := make(map[int]int)
	for i := 0; i < len(graph); i++ {
		if dfs(graph, i, seen) == false {
			return false
		}
	}
	return true
}

func dfs(graph [][]int, curIdx int, seen map[int]int) bool {
	if _, x := seen[curIdx]; x {
		if seen[curIdx] == 1 {
			// if seen[curIdx] == 1, it meas this node is being visiting and here is a cycle
			return false
		} else if seen[curIdx] == 2 {
			// if seen[curIdx] == 2, it meas this node has been visited and no cycle is detected
			return true
		}
	}
	// mark the curidx as 'visiting'
	seen[curIdx] = 1
	children := graph[curIdx]
	for i := 0; i < len(children); i++ {
		if dfs(graph, children[i], seen) == false {
			return false
		}
	}
	// mark the curidx as 'visited'
	seen[curIdx] = 2
	return true
}

func main() {
	fmt.Println(canFinish(6, [][]int{
		{4, 3}, {1, 0}, {5, 2}, {5, 4}, {5, 1}, {2, 3},
	}))

	fmt.Println(canFinish(6, [][]int{
		{4, 3}, {1, 0}, {5, 2}, {5, 4}, {5, 1}, {2, 3}, {3, 5},
	}))
}
